File: src/lib/engine/tracker.ts
import { STEP_SECONDS } from "./mock-candles";
import type { MarketType } from "./binance";
import type { TradingIntent } from "./intent";
import type { TokenTimeframe } from "./mock-candles";
import type { SetupType, TradeDirection } from "./quant";
import type { Candle } from "./types";

/** "active" is the only open state — following a signal means you've already entered. */
export type TrackedSignalStatus = "active" | "target1-hit" | "target2-hit" | "stopped-out";

export function isTerminalStatus(status: TrackedSignalStatus): boolean {
  return status !== "active";
}

export interface TrackedSignal {
  id: string;
  symbol: string;
  intent: TradingIntent;
  direction: Exclude<TradeDirection, "none">;
  setupType: SetupType;
  /** Execution timeframe the plan was built on. */
  timeframe: TokenTimeframe;
  /** Binance market the signal priced against; absent on older records (spot). */
  market?: MarketType;
  /** The engine's ideal entry zone at follow time, kept for reference against the actual entry price. */
  entryLow: number;
  entryHigh: number;
  /** The price the user confirmed they actually entered at. */
  entryPrice: number;
  stop: number;
  target1: number;
  target2: number;
  confidenceAtFollow: number;
  followedAt: string;
  status: TrackedSignalStatus;
  closePrice?: number;
  closedAt?: string;
  resultR?: number;
}

function round(value: number, digits = 2): number {
  const scale = 10 ** digits;
  return Math.round(value * scale) / scale;
}

/**
 * Advances one tracked signal against a fresh price tick. Returns a patch to
 * merge, or null when nothing changed. First-touch-wins, same walk-forward
 * philosophy as `runBacktest` in quant.ts: whichever of stop/target1/target2
 * price reaches first closes the trade.
 *
 * Known limitation: this only sees polled last price, not intrabar highs/
 * lows, so a wick through a level between polls can be missed — an accepted
 * v1 gap (see the "Ideal entry range + signal tracker" plan).
 */
export function evaluateTrackedSignal(
  signal: TrackedSignal,
  latestPrice: number,
  nowIso: string,
): Partial<TrackedSignal> | null {
  if (!Number.isFinite(latestPrice) || latestPrice <= 0) return null;
  if (isTerminalStatus(signal.status)) return null;

  const long = signal.direction === "long";
  const riskPerUnit = Math.abs(signal.entryPrice - signal.stop);
  const resultR = (exit: number) =>
    riskPerUnit > 0
      ? round((long ? exit - signal.entryPrice : signal.entryPrice - exit) / riskPerUnit, 2)
      : 0;

  const stopHit = long ? latestPrice <= signal.stop : latestPrice >= signal.stop;
  if (stopHit) {
    return {
      status: "stopped-out",
      closePrice: latestPrice,
      closedAt: nowIso,
      resultR: resultR(signal.stop),
    };
  }
  const target2Hit = long ? latestPrice >= signal.target2 : latestPrice <= signal.target2;
  if (target2Hit) {
    return {
      status: "target2-hit",
      closePrice: latestPrice,
      closedAt: nowIso,
      resultR: resultR(signal.target2),
    };
  }
  const target1Hit = long ? latestPrice >= signal.target1 : latestPrice <= signal.target1;
  if (target1Hit) {
    return {
      status: "target1-hit",
      closePrice: latestPrice,
      closedAt: nowIso,
      resultR: resultR(signal.target1),
    };
  }
  return null;
}

export interface ExitLevels {
  direction: "long" | "short";
  entry: number;
  stop: number;
  target1: number;
  target2: number;
}

export interface ExitEvent {
  status: "stopped-out" | "target1-hit" | "target2-hit";
  /** The level that produced the exit (stop or target price). */
  exitLevel: number;
  /** Open time (sec) of the bar that touched the level. */
  barTime: number;
  resultR: number;
}

/**
 * Walks closed bars first-touch-wins using intrabar highs/lows. Within a
 * single bar the stop is checked first (conservative, same precedence as
 * `runBacktest`), then target 2, then target 1.
 */
export function walkExitLevels(levels: ExitLevels, bars: Candle[]): ExitEvent | null {
  const long = levels.direction === "long";
  const riskPerUnit = Math.abs(levels.entry - levels.stop);
  const resultR = (exit: number) =>
    riskPerUnit > 0
      ? round((long ? exit - levels.entry : levels.entry - exit) / riskPerUnit, 2)
      : 0;

  for (const bar of bars) {
    if (long ? bar.low <= levels.stop : bar.high >= levels.stop) {
      return {
        status: "stopped-out",
        exitLevel: levels.stop,
        barTime: bar.time,
        resultR: resultR(levels.stop),
      };
    }
    if (long ? bar.high >= levels.target2 : bar.low <= levels.target2) {
      return {
        status: "target2-hit",
        exitLevel: levels.target2,
        barTime: bar.time,
        resultR: resultR(levels.target2),
      };
    }
    if (long ? bar.high >= levels.target1 : bar.low <= levels.target1) {
      return {
        status: "target1-hit",
        exitLevel: levels.target1,
        barTime: bar.time,
        resultR: resultR(levels.target1),
      };
    }
  }
  return null;
}

/**
 * Catch-up settlement against closed klines: unlike `evaluateTrackedSignal`
 * (polled last price), this sees intrabar highs/lows, so wicks through a
 * level between polls are no longer missed. Only bars that opened after the
 * follow time count — the bar in progress at follow time partially predates
 * the entry.
 */
export function settleTrackedSignalWithCandles(
  signal: TrackedSignal,
  closedBars: Candle[],
): Partial<TrackedSignal> | null {
  if (isTerminalStatus(signal.status)) return null;
  const followedAtSec = Date.parse(signal.followedAt) / 1000;
  if (!Number.isFinite(followedAtSec)) return null;

  const bars = closedBars.filter((c) => c.time > followedAtSec);
  const exit = walkExitLevels(
    {
      direction: signal.direction,
      entry: signal.entryPrice,
      stop: signal.stop,
      target1: signal.target1,
      target2: signal.target2,
    },
    bars,
  );
  if (!exit) return null;

  return {
    status: exit.status,
    closePrice: exit.exitLevel,
    closedAt: new Date((exit.barTime + STEP_SECONDS[signal.timeframe]) * 1000).toISOString(),
    resultR: exit.resultR,
  };
}

export interface TrackedSignalSummary {
  total: number;
  open: number;
  closed: number;
  wins: number;
  losses: number;
  winRate: number;
  averageR: number;
  /** True when the closed sample is too small (<5) for winRate/averageR to be meaningful. */
  lowSample: boolean;
}

const MIN_RELIABLE_TRACKED_TRADES = 5;

/** Aggregate stats over followed signals. */
export function summarizeTrackedSignals(signals: TrackedSignal[]): TrackedSignalSummary {
  const closed = signals.filter((s) => isTerminalStatus(s.status));
  const wins = closed.filter((s) => s.status === "target1-hit" || s.status === "target2-hit");
  const losses = closed.filter((s) => s.status === "stopped-out");
  const rValues = closed.map((s) => s.resultR ?? 0);
  const averageR = rValues.length ? rValues.reduce((sum, r) => sum + r, 0) / rValues.length : 0;

  return {
    total: signals.length,
    open: signals.filter((s) => !isTerminalStatus(s.status)).length,
    closed: closed.length,
    wins: wins.length,
    losses: losses.length,
    winRate: closed.length ? round((wins.length / closed.length) * 100, 1) : 0,
    averageR: round(averageR, 2),
    lowSample: closed.length > 0 && closed.length < MIN_RELIABLE_TRACKED_TRADES,
  };
}

File: engine/smc/tracker.py
"""Followed-signal tracking + the shared exit-level walk.

``walk_exit_levels`` is the one first-touch-wins convention every settlement
path uses: within a single bar the stop is checked first (conservative), then
target 2, then target 1. Ported verbatim from the TS engine's ``tracker.ts``.
"""

from __future__ import annotations

import math
from dataclasses import dataclass
from typing import Literal

from smc.hysteresis import iso_from_ms, parse_iso_ms
from smc.intent import TradingIntent
from smc.mock_candles import STEP_SECONDS, TokenTimeframe
from smc.quant import SetupType
from smc.types import Candle, MarketType

# "active" is the only open state — following a signal means you've entered.
TrackedSignalStatus = Literal["active", "target1-hit", "target2-hit", "stopped-out"]


def is_terminal_status(status: TrackedSignalStatus) -> bool:
    return status != "active"


@dataclass(slots=True)
class TrackedSignal:
    id: str
    symbol: str
    intent: TradingIntent
    direction: Literal["long", "short"]
    setup_type: SetupType
    # Execution timeframe the plan was built on.
    timeframe: TokenTimeframe
    # The engine's ideal entry zone at follow time, kept for reference.
    entry_low: float
    entry_high: float
    # The price the user confirmed they actually entered at.
    entry_price: float
    stop: float
    target1: float
    target2: float
    confidence_at_follow: float
    followed_at: str
    status: TrackedSignalStatus
    # Binance market the signal priced against; None on older records (spot).
    market: MarketType | None = None
    close_price: float | None = None
    closed_at: str | None = None
    result_r: float | None = None
    # Provenance — engine version / config / commit at follow time.
    engine_version: str | None = None
    config_hash: str | None = None
    git_sha: str | None = None


@dataclass(slots=True)
class TrackedPatch:
    """The fields a settlement pass may change on a tracked signal."""

    status: TrackedSignalStatus
    close_price: float
    closed_at: str
    result_r: float


def _round(value: float, digits: int = 2) -> float:
    scale = 10.0**digits
    return math.floor(value * scale + 0.5) / scale


def evaluate_tracked_signal(
    signal: TrackedSignal, latest_price: float, now_iso: str
) -> TrackedPatch | None:
    """Advances one tracked signal against a fresh price tick. Returns a patch
    to merge, or None when nothing changed. Known limitation: only sees polled
    last price, not intrabar highs/lows — an accepted v1 gap."""
    if not math.isfinite(latest_price) or latest_price <= 0:
        return None
    if is_terminal_status(signal.status):
        return None

    long = signal.direction == "long"
    risk_per_unit = abs(signal.entry_price - signal.stop)

    def result_r(exit_: float) -> float:
        if risk_per_unit <= 0:
            return 0
        gain = exit_ - signal.entry_price if long else signal.entry_price - exit_
        return _round(gain / risk_per_unit, 2)

    stop_hit = latest_price <= signal.stop if long else latest_price >= signal.stop
    if stop_hit:
        return TrackedPatch(
            status="stopped-out",
            close_price=latest_price,
            closed_at=now_iso,
            result_r=result_r(signal.stop),
        )
    target2_hit = latest_price >= signal.target2 if long else latest_price <= signal.target2
    if target2_hit:
        return TrackedPatch(
            status="target2-hit",
            close_price=latest_price,
            closed_at=now_iso,
            result_r=result_r(signal.target2),
        )
    target1_hit = latest_price >= signal.target1 if long else latest_price <= signal.target1
    if target1_hit:
        return TrackedPatch(
            status="target1-hit",
            close_price=latest_price,
            closed_at=now_iso,
            result_r=result_r(signal.target1),
        )
    return None


@dataclass(slots=True)
class ExitLevels:
    direction: Literal["long", "short"]
    entry: float
    stop: float
    target1: float
    target2: float


@dataclass(slots=True)
class ExitEvent:
    status: Literal["stopped-out", "target1-hit", "target2-hit"]
    # The level that produced the exit (stop or target price).
    exit_level: float
    # Open time (sec) of the bar that touched the level.
    bar_time: int
    result_r: float


def walk_exit_levels(levels: ExitLevels, bars: list[Candle]) -> ExitEvent | None:
    """Walks closed bars first-touch-wins using intrabar highs/lows. Within a
    single bar the stop is checked first, then target 2, then target 1."""
    long = levels.direction == "long"
    risk_per_unit = abs(levels.entry - levels.stop)

    def result_r(exit_: float) -> float:
        if risk_per_unit <= 0:
            return 0
        gain = exit_ - levels.entry if long else levels.entry - exit_
        return _round(gain / risk_per_unit, 2)

    for bar in bars:
        if bar.low <= levels.stop if long else bar.high >= levels.stop:
            return ExitEvent(
                status="stopped-out",
                exit_level=levels.stop,
                bar_time=bar.time,
                result_r=result_r(levels.stop),
            )
        if bar.high >= levels.target2 if long else bar.low <= levels.target2:
            return ExitEvent(
                status="target2-hit",
                exit_level=levels.target2,
                bar_time=bar.time,
                result_r=result_r(levels.target2),
            )
        if bar.high >= levels.target1 if long else bar.low <= levels.target1:
            return ExitEvent(
                status="target1-hit",
                exit_level=levels.target1,
                bar_time=bar.time,
                result_r=result_r(levels.target1),
            )
    return None


def settle_tracked_signal_with_candles(
    signal: TrackedSignal, closed_bars: list[Candle]
) -> TrackedPatch | None:
    """Catch-up settlement against closed klines: sees intrabar highs/lows, so
    wicks through a level between polls are no longer missed. Only bars that
    opened after the follow time count — the bar in progress at follow time
    partially predates the entry."""
    if is_terminal_status(signal.status):
        return None
    followed_at_sec = parse_iso_ms(signal.followed_at) / 1000
    if not math.isfinite(followed_at_sec):
        return None

    bars = [c for c in closed_bars if c.time > followed_at_sec]
    exit_ = walk_exit_levels(
        ExitLevels(
            direction=signal.direction,
            entry=signal.entry_price,
            stop=signal.stop,
            target1=signal.target1,
            target2=signal.target2,
        ),
        bars,
    )
    if exit_ is None:
        return None

    return TrackedPatch(
        status=exit_.status,
        close_price=exit_.exit_level,
        closed_at=iso_from_ms((exit_.bar_time + STEP_SECONDS[signal.timeframe]) * 1000),
        result_r=exit_.result_r,
    )


@dataclass(slots=True)
class TrackedSignalSummary:
    total: int
    open: int
    closed: int
    wins: int
    losses: int
    win_rate: float
    average_r: float
    # True when the closed sample is too small (<5) to be meaningful.
    low_sample: bool


_MIN_RELIABLE_TRACKED_TRADES = 5


def summarize_tracked_signals(signals: list[TrackedSignal]) -> TrackedSignalSummary:
    """Aggregate stats over followed signals."""
    closed = [s for s in signals if is_terminal_status(s.status)]
    wins = [s for s in closed if s.status in ("target1-hit", "target2-hit")]
    losses = [s for s in closed if s.status == "stopped-out"]
    r_values = [s.result_r if s.result_r is not None else 0 for s in closed]
    average_r = sum(r_values) / len(r_values) if r_values else 0

    return TrackedSignalSummary(
        total=len(signals),
        open=len(signals) - len(closed),
        closed=len(closed),
        wins=len(wins),
        losses=len(losses),
        win_rate=_round(len(wins) / len(closed) * 100, 1) if closed else 0,
        average_r=_round(average_r, 2),
        low_sample=0 < len(closed) < _MIN_RELIABLE_TRACKED_TRADES,
    )
